### PR TITLE
feature: address typescript issues

### PR DESCRIPTION
**Pull request by murbanowicz**
This PR moves whole codebase to TypeScript and enables build for ESM and UMD, including the types.

I tested it quickly by installing the package with 1. npm pack 2. adding it to package json with 'file:/...' and it works nicely:
image

It depends on types from kinde-pkce-js which should be adjusted and made more precise, but then the types in this package will auto update after updating this dependency as they rely on the types of the KindeClient exported from that library.

You may want to adjust it, but I believe majority of work is done.

**TERSER is not added to the build**. It was breaking the build completely to me and I had to remove it at this stage. 
I am not that familiar with Rollup to get it working in reasonable timeframe.
Long story short - after adding terser, it starts to break with weird CJS error on config, while without it, it works fine.

Feel free to contact me on Slack if you need any more help with this and obviously to edit this by adding other required stuff.

### CODE REVIEW

File: rollup.config.js [hunk 1]
```diff
@@ -1,35 +1,41 @@
...
 const require = createRequire(import.meta.url);
 const pkg = require('./package.json');
-import terser from '@rollup/plugin-terser';
-import babel from '@rollup/plugin-babel';
-import dts from 'rollup-plugin-dts';
 
+const commonPlugins = [
+  typescript({ tsconfig: './tsconfig.json' }),
+  commonjs(),
+    // terser()
```
Comment (DaveOrDead) [inline rollup.config.js:15]: Hey @murbanowicz thanks for the PR - looking awesome 🚀 

Just wondering if there was a reason for removing minification?

Comment (murbanowicz) [reply rollup.config.js:15]: Hi Mate! Look at the bottom part of the PR description. I would ask you to help on this and manage rollup. I've spent quite a lot of time on this already which makes it pointless on my side.

Comment (DaveOrDead) [reply rollup.config.js:15]: @murbanowicz my bad, missed that part of the description. Looking into it now :)

We've updated the types in the PKCE vanilla JS lib this depends on now also which may make some of the typings here better

File: src/state/reducer.ts [hunk 2]
```diff
@@ -1,27 +1,30 @@
...
-const onComplete = (state, action) =>
-  state.user && state.user.updated_at === action.user && action.user.updated_at
+const onComplete = (state: State, action: Partial<State>): State =>
+  state.user?.updated_at === action.user?.updated_at
     ? state
     : {
         ...state,
         isAuthenticated: !!action.user,
         user: action.user
       };
```
Comment (DaveOrDead) [inline src/state/reducer.ts:19]: I've realised `onComplete` actually isn't used anywhere so this whole function can be removed

File: src/state/reducer.ts [hunk 3]
```diff
@@ -35,10 +38,8 @@ const reducerMap = {
   ERROR: onError
```
Comment (DaveOrDead) [inline src/state/reducer.ts:38]: As we're removing the `onComplete` function, these two lines can also be removed
```
ON_REDIRECT_COMPLETE: onComplete,
GET_ACCESS_TOKEN_COMPLETE: onComplete,
```

File: src/state/types.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+export type KindeUser = {
+  given_name: string | null;
+  id: string | null;
+  family_name: string | null;
+  email: string | null;
+  updated_at: number;
```
Comment (DaveOrDead) [inline src/state/types.ts:6]: `updated_at` isn't actually exposed anywhere in the user object currently, so we can remove this for now

File: rollup.config.js (hunk 1 above)
Comment (DaveOrDead) [reply rollup.config.js:15]: It seems that this issue is with upgrading terser from `0.1.0` to `0.2.0` as per [this issue](https://github.com/rollup/plugins/issues/1366) 

If you revert the change in `package.json` back to `"@rollup/plugin-terser": "^0.1.0"` then it all builds fine :) 

Hopefully they resolve soon so we can use the latest, but I'm ok with it being on 0.1.0 for now

**Review by DaveOrDead — CHANGES_REQUESTED** [review]

**Comment by murbanowicz** [comment]
I've addressed all the remarks. I had to remove type:module from package.json to fix ESLint, then change rollup config extension to mjs to get it working without it, but now it all build etc. 
I hope it's fine :)

**Review by DaveOrDead — APPROVED** [review]
